proxy select table throw NullPointerException!

use:sharding-proxy-4.0.0

stdout.log :
`
[ERROR] 20:47:17.372 [ShardingSphere-Command-13] o.a.s.s.f.c.CommandExecutorTask - Exception occur: 
java.lang.NullPointerException: null
        at org.apache.shardingsphere.shardingproxy.backend.response.query.QueryHeader.<init>(QueryHeader.java:67)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.execute.callback.ProxySQLExecuteCallback.getQueryHeaders(ProxySQLExecuteCallback.java:89)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.execute.callback.ProxySQLExecuteCallback.executeSQL(ProxySQLExecuteCallback.java:81)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.execute.callback.ProxySQLExecuteCallback.executeSQL(ProxySQLExecuteCallback.java:73)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.execute.callback.ProxySQLExecuteCallback.executeSQL(ProxySQLExecuteCallback.java:45)
        at org.apache.shardingsphere.core.execute.sql.execute.SQLExecuteCallback.execute0(SQLExecuteCallback.java:79)
        at org.apache.shardingsphere.core.execute.sql.execute.SQLExecuteCallback.execute(SQLExecuteCallback.java:61)
        at org.apache.shardingsphere.core.execute.engine.ShardingExecuteEngine.syncGroupExecute(ShardingExecuteEngine.java:125)
        at org.apache.shardingsphere.core.execute.engine.ShardingExecuteEngine.parallelExecute(ShardingExecuteEngine.java:102)
        at org.apache.shardingsphere.core.execute.engine.ShardingExecuteEngine.groupExecute(ShardingExecuteEngine.java:83)
        at org.apache.shardingsphere.core.execute.sql.execute.SQLExecuteTemplate.executeGroup(SQLExecuteTemplate.java:73)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.execute.JDBCExecuteEngine.execute(JDBCExecuteEngine.java:82)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.JDBCDatabaseCommunicationEngine.execute(JDBCDatabaseCommunicationEngine.java:99)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.JDBCDatabaseCommunicationEngine.execute(JDBCDatabaseCommunicationEngine.java:84)
        at org.apache.shardingsphere.shardingproxy.backend.text.query.QueryBackendHandler.execute(QueryBackendHandler.java:54)
        at org.apache.shardingsphere.shardingproxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:73)
        at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:92)
        at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:72)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

`

proxy log:
`
2020-01-08 20:55:48,644 [ShardingSphere-Command-2] INFO [ShardingSphere-SQL] Rule Type: sharding
2020-01-08 20:55:48,645 [ShardingSphere-Command-2] INFO [ShardingSphere-SQL] Logic SQL: /* ApplicationName=DBeaver 6.1.2 - Main */ SET SQL_SELECT_LIMIT=200
2020-01-08 20:55:48,645 [ShardingSphere-Command-2] INFO [ShardingSphere-SQL] SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.SetStatement@178abd17, tablesContext=TablesContext(tables=[], schema=Optional.absent()))
2020-01-08 20:55:48,645 [ShardingSphere-Command-2] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SET SQL_SELECT_LIMIT=200
2020-01-08 20:55:48,658 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Rule Type: sharding
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Logic SQL: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] SQLStatement: SelectSQLStatementContext(super=CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dml.SelectStatement@77adb0a6, tablesContext=TablesContext(tables=[Table(name=t_role_assets, alias=Optional.absent())], schema=Optional.absent())), projectionsContext=ProjectionsContext(startIndex=50, stopIndex=50, distinctRow=false, projections=[ShorthandProjection(owner=Optional.absent())], columnLabels=[roleid, dailydiamondconsume, dailyrecharge, diamond, energy, gold, guildgold, roleexp, totaldiamond, totaldropgold, totalexp, totalrecharge, vippoint, energytime]), groupByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.groupby.GroupByContext@5dd9ba9a, orderByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.orderby.OrderByContext@8965fd, paginationContext=org.apache.shardingsphere.sql.parser.relation.segment.select.pagination.PaginationContext@1934d0c0, containsSubquery=false)
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets0 limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets1 limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets2 limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub1 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets0 limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub1 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets1 limit 10
2020-01-08 20:55:48,659 [ShardingSphere-Command-14] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub1 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SELECT * from t_role_assets2 limit 10
2020-01-08 20:55:48,670 [ShardingSphere-Command-4] INFO [ShardingSphere-SQL] Rule Type: sharding
2020-01-08 20:55:48,670 [ShardingSphere-Command-4] INFO [ShardingSphere-SQL] Logic SQL: /* ApplicationName=DBeaver 6.1.2 - Main */ SET SQL_SELECT_LIMIT=DEFAULT
2020-01-08 20:55:48,670 [ShardingSphere-Command-4] INFO [ShardingSphere-SQL] SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.SetStatement@6d0df86d, tablesContext=TablesContext(tables=[], schema=Optional.absent()))
2020-01-08 20:55:48,670 [ShardingSphere-Command-4] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SET SQL_SELECT_LIMIT=DEFAULT
2020-01-08 20:55:48,675 [ShardingSphere-Command-5] INFO [ShardingSphere-SQL] Rule Type: sharding
2020-01-08 20:55:48,675 [ShardingSphere-Command-5] INFO [ShardingSphere-SQL] Logic SQL: /* ApplicationName=DBeaver 6.1.2 - Main */ SHOW WARNINGS
2020-01-08 20:55:48,675 [ShardingSphere-Command-5] INFO [ShardingSphere-SQL] SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.dialect.mysql.ShowOtherStatement@50305fb5, tablesContext=TablesContext(tables=[], schema=Optional.absent()))
2020-01-08 20:55:48,675 [ShardingSphere-Command-5] INFO [ShardingSphere-SQL] Actual SQL: r_game_sub0 ::: /* ApplicationName=DBeaver 6.1.2 - Main */ SHOW WARNINGS


`

--- a/sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/rule/ShardingRule.java
+++ b/sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/rule/ShardingRule.java
@@ -382,7 +382,7 @@ public class ShardingRule implements BaseRule {
     }
     
     /**
-     * Get logic table names base on actual table name.
+     * Get logic table names based on actual table name.
      *
      * @param actualTableName actual table name
      * @return logic table name

--- a/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
+++ b/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
@@ -71,7 +71,7 @@ public final class QueryHeader {
         String actualTableName = resultSetMetaData.getTableName(columnIndex);
         if (null != actualTableName && logicSchema instanceof ShardingSchema) {
             Collection<String> logicTableNames = logicSchema.getShardingRule().getLogicTableNames(actualTableName);
-            table = logicTableNames.isEmpty()? "" : logicTableNames.iterator().next();
+            table = logicTableNames.isEmpty() ? "" : logicTableNames.iterator().next();
             TableMetaData tableMetaData = logicSchema.getMetaData().getTables().get(table);
             primaryKey = null != tableMetaData && tableMetaData.getColumns().get(resultSetMetaData.getColumnName(columnIndex).toLowerCase())
                     .isPrimaryKey();

--- a/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
+++ b/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
@@ -25,6 +25,7 @@ import org.apache.shardingsphere.underlying.common.metadata.table.TableMetaData;
 
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
+import java.util.Collection;
 
 /**
  * Query header.
@@ -69,7 +70,8 @@ public final class QueryHeader {
         autoIncrement = resultSetMetaData.isAutoIncrement(columnIndex);
         String actualTableName = resultSetMetaData.getTableName(columnIndex);
         if (null != actualTableName && logicSchema instanceof ShardingSchema) {
-            table = logicSchema.getShardingRule().getLogicTableNames(actualTableName).iterator().next();
+            Collection<String> logicTableNames = logicSchema.getShardingRule().getLogicTableNames(actualTableName);
+            table = logicTableNames.isEmpty()? null : logicTableNames.iterator().next();
             TableMetaData tableMetaData = logicSchema.getMetaData().getTables().get(table);
             primaryKey = null != tableMetaData && tableMetaData.getColumns().get(resultSetMetaData.getColumnName(columnIndex).toLowerCase())
                     .isPrimaryKey();

--- a/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
+++ b/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
@@ -71,7 +71,7 @@ public final class QueryHeader {
         String actualTableName = resultSetMetaData.getTableName(columnIndex);
         if (null != actualTableName && logicSchema instanceof ShardingSchema) {
             Collection<String> logicTableNames = logicSchema.getShardingRule().getLogicTableNames(actualTableName);
-            table = logicTableNames.isEmpty()? null : logicTableNames.iterator().next();
+            table = logicTableNames.isEmpty()? "" : logicTableNames.iterator().next();
             TableMetaData tableMetaData = logicSchema.getMetaData().getTables().get(table);
             primaryKey = null != tableMetaData && tableMetaData.getColumns().get(resultSetMetaData.getColumnName(columnIndex).toLowerCase())
                     .isPrimaryKey();

--- a/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
+++ b/sharding-proxy/sharding-proxy-backend/src/main/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeader.java
@@ -21,10 +21,10 @@ import lombok.AllArgsConstructor;
 import lombok.Getter;
 import org.apache.shardingsphere.shardingproxy.backend.schema.LogicSchema;
 import org.apache.shardingsphere.shardingproxy.backend.schema.impl.ShardingSchema;
+import org.apache.shardingsphere.underlying.common.metadata.table.TableMetaData;
 
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
-import java.util.Collection;
 
 /**
  * Query header.
@@ -58,29 +58,25 @@ public final class QueryHeader {
     private final boolean autoIncrement;
     
     public QueryHeader(final ResultSetMetaData resultSetMetaData, final LogicSchema logicSchema, final int columnIndex) throws SQLException {
-        this.schema = logicSchema.getName();
-        String actualTableName = null == resultSetMetaData.getTableName(columnIndex) ? "" : resultSetMetaData.getTableName(columnIndex);
-        if (logicSchema instanceof ShardingSchema) {
-            Collection<String> tableNames = logicSchema.getShardingRule().getLogicTableNames(actualTableName);
-            this.table = tableNames.isEmpty() ? "" : tableNames.iterator().next();
-            if (logicSchema.getMetaData().getTables().containsTable(actualTableName)) {
-                this.primaryKey = logicSchema.getMetaData().getTables().get(actualTableName).getColumns()
-                        .get(resultSetMetaData.getColumnName(columnIndex).toLowerCase()).isPrimaryKey();
-            } else {
-                this.primaryKey = false;
-            }
+        schema = logicSchema.getName();
+        columnLabel = resultSetMetaData.getColumnLabel(columnIndex);
+        columnName = resultSetMetaData.getColumnName(columnIndex);
+        columnLength = resultSetMetaData.getColumnDisplaySize(columnIndex);
+        columnType = resultSetMetaData.getColumnType(columnIndex);
+        decimals = resultSetMetaData.getScale(columnIndex);
+        signed = resultSetMetaData.isSigned(columnIndex);
+        notNull = resultSetMetaData.isNullable(columnIndex) == ResultSetMetaData.columnNoNulls;
+        autoIncrement = resultSetMetaData.isAutoIncrement(columnIndex);
+        String actualTableName = resultSetMetaData.getTableName(columnIndex);
+        if (null != actualTableName && logicSchema instanceof ShardingSchema) {
+            table = logicSchema.getShardingRule().getLogicTableNames(actualTableName).iterator().next();
+            TableMetaData tableMetaData = logicSchema.getMetaData().getTables().get(table);
+            primaryKey = null != tableMetaData && tableMetaData.getColumns().get(resultSetMetaData.getColumnName(columnIndex).toLowerCase())
+                    .isPrimaryKey();
         } else {
-            this.table = actualTableName;
-            this.primaryKey = false;
+            table = actualTableName;
+            primaryKey = false;
         }
-        this.columnLabel = resultSetMetaData.getColumnLabel(columnIndex);
-        this.columnName = resultSetMetaData.getColumnName(columnIndex);
-        this.columnLength = resultSetMetaData.getColumnDisplaySize(columnIndex);
-        this.columnType = resultSetMetaData.getColumnType(columnIndex);
-        this.decimals = resultSetMetaData.getScale(columnIndex);
-        this.signed = resultSetMetaData.isSigned(columnIndex);
-        this.notNull = resultSetMetaData.isNullable(columnIndex) == ResultSetMetaData.columnNoNulls;
-        this.autoIncrement = resultSetMetaData.isAutoIncrement(columnIndex);
     }
     
     /**

--- a/sharding-proxy/sharding-proxy-backend/src/test/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeaderTest.java
+++ b/sharding-proxy/sharding-proxy-backend/src/test/java/org/apache/shardingsphere/shardingproxy/backend/response/query/QueryHeaderTest.java
@@ -111,8 +111,7 @@ public final class QueryHeaderTest {
         ShardingSchema result = mock(ShardingSchema.class);
         ColumnMetaData columnMetaData = new ColumnMetaData("order_id", "int", true);
         TableMetas tableMetas = mock(TableMetas.class);
-        when(tableMetas.get("t_order")).thenReturn(new TableMetaData(Collections.singletonList(columnMetaData), Collections.singletonList("order_id")));
-        when(tableMetas.containsTable("t_order")).thenReturn(true);
+        when(tableMetas.get("t_logic_order")).thenReturn(new TableMetaData(Collections.singletonList(columnMetaData), Collections.singletonList("order_id")));
         ShardingSphereMetaData metaData = mock(ShardingSphereMetaData.class);
         when(metaData.getTables()).thenReturn(tableMetas);
         DataSourceMetas dataSourceMetas = mock(DataSourceMetas.class);

